### ISSUE TITLE
session.find on non primary key columns. Suitable Alternative

### ISSUE THREAD

**Issue by coder-github578** [hibernate/hibernate-reactive#2024]

**Comment by coder-github578** [comment]
Is there any way to do session.find on non-primary key column

**Comment by gavinking** [comment]
Yes, using `Session.find(Class<T> entityClass, Identifier<T> naturalId);`

**Comment by coder-github578** [comment]
Thanks Gavin for the quick response. So do i have to mark @NaturalId attribute to the columns in the entity model whenever i want to make a search.

**Comment by coder-github578** [comment]
Thanks that worked

**Comment by coder-github578** [comment]
Can this scenario work if the column is not unique, like i need to query to status

**Comment by coder-github578** [comment]
createSelectionQuery doesnt consider the relationship with the other table. i have to explicitly mention Entity1 and Entity2 whereas in session.find it automatically does the join based on the relationship specified. Correct me if i am wrong, i am very new to this hibernate reactive

**Comment by DavideD** [comment]
> Correct me if i am wrong

You are wrong :)

`.createSelectionQuery` works the same as `.find` in regards to loading associations. 
Hibernate ORM, or Hibernate Reactive, decides which association to load based on the entity mapping, or some options that you can add to the query.

You will find all the information related to the fetching of associations in the Hibernate ORM documentation: https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#fetching

The only difference it that Hibernate Reactive cannot load lazy associations transparently, they need to be explicitly fetched: https://hibernate.org/reactive/documentation/2.4/reference/html_single/#_fetching_lazy_associations

**Comment by coder-github578** [comment]
Got it. This may not be relevant but in hibernate reactive, is there any method to make projection with more than 15 columns to DTO. I am currently planning to use Mapstruct to map Entity to DTO. Any kind of solution will be appreciated

**Comment by DavideD** [comment]
I don't think there is. You can return the result as `Tuple.class`, though. This is the conversation we had in the past about it: https://github.com/hibernate/hibernate-reactive/issues/1209#issuecomment-1049753624

**Comment by DavideD** [comment]
@gavinking I've just noticed that Hibernate ORM has a [setTupleTransformer](https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#sql-dto-query), but Hibernate Reactive doesn't have it in the API. I guess we should add it, isn't it?

**Comment by coder-github578** [comment]
Thanks will try something on these lines

**Comment by gavinking** [comment]
> is there any method to make projection with more than 15 columns to DTO.

What is wrong with `select new DTO(x, y, z) ...`. Actually in H6, the `new DTO` but is not even required if you specify a result type explicitly.

**Comment by gavinking** [comment]
I mean this is supposed to just work:

```java
session.createSelectionQuery("select title, isbn, id from books", Dto.class ).getResultList()
```

It certainly works in regular Hibernate.

**Comment by gavinking** [comment]
@DavideD 

> Hibernate ORM has a [setTupleTransformer](https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#sql-dto-query), but Hibernate Reactive doesn't have it in the API. I guess we should add it, isn't it?

Meh, the only person I know who likes that is @maxandersen. I have never seen much enthusiasm for it from users. How is it better than just using `java.util.Stream.map()`?

**Comment by DavideD** [comment]
> I mean this is supposed to just work:
> 
> ```java
> session.createSelectionQuery("select title, isbn, id from books", Dto.class ).getResultList()
> ```
> 
> It certainly works in regular Hibernate.

I didn't know we could do this. I will have to add a test in Reactive.
Anyway, it seems users have enough way to solve this issue.
Thanks

**Comment by coder-github578** [comment]
In My Case the number of columns that i have is more than 15-20 columns and writing each and every column is little tedious every time, plus i have one-to-one and one-to-many relationship. All the examples i have seen so far talk about single table and few columns. Hence i was looking for some kind of auto mapping approach from hibernate reactive

**Comment by maxandersen** [comment]
> Meh, the only person I know who likes that is @maxandersen. I have never seen much enthusiasm for it from users. How is it better than just using `java.util.Stream.map()`?

Just liked it because at the time you couldn't do what you just showed - map() works but I'm gonna assume that is requiring allocating the full item twice at minum rather than just copy raw result to a result row?

**Comment by gavinking** [comment]
>  I'm gonna assume that is requiring allocating the full item twice at minum rather than just copy raw result to a result row

It's going to allocate an array either way.
